hail/python/test/hailtop/hailctl/batch/test_submit.py

```diff
@@ -0,0 +1,139 @@
+        assert res.exit_code == 0, repr((res.output, res.stdout, res.stderr, res.exception))
+
+
+@pytest.mark.timeout(5 * 60)  # image pulling is very slow
+def test_file_mount_in_child_dir_to_root_dir(runner: CliRunner):
+    with tempfile.TemporaryDirectory() as dir:
+        os.chdir(dir)
+        write_hello(f'{dir}/child/hello.txt')
+        write_script(dir, '/hello.txt')
+        res = runner.invoke(
+            cli.app, ['submit', '--wait', '--files', 'child/hello.txt:/', 'test_job.py'], catch_exceptions=False
+        )
+        assert res.exit_code == 0, repr((res.output, res.stdout, res.stderr, res.exception))
+
+
+@pytest.mark.timeout(5 * 60)  # image pulling is very slow
+def test_mount_multiple_files(runner: CliRunner):
+    with tempfile.TemporaryDirectory() as dir:
+        os.chdir(dir)
+        write_hello(f'{dir}/child/hello1.txt')
+        write_hello(f'{dir}/child/hello2.txt')
+        write_script(dir, '/hello1.txt')
+        res = runner.invoke(
+            cli.app,
+            [
+                'submit',
+                '--wait',
+                '--files',
+                'child/hello1.txt:/',
+                '--files',
+                'child/hello2.txt:/',
+                'test_job.py',
+            ],
+            catch_exceptions=False,
+        )
+        assert res.exit_code == 0, repr((res.output, res.stdout, res.stderr, res.exception))
+
+
+@pytest.mark.timeout(5 * 60)  # image pulling is very slow
+def test_dir_mount_in_child_dir_to_child_dir(runner: CliRunner):
+    with tempfile.TemporaryDirectory() as dir:
+        os.chdir(dir)
+        write_hello(f'{dir}/child/hello1.txt')
+        write_hello(f'{dir}/child/hello2.txt')
+        write_script(dir, '/child/hello1.txt')
+        res = runner.invoke(
+            cli.app, ['submit', '--wait', '--files', 'child/:/child/', 'test_job.py'], catch_exceptions=False
+        )
+        assert res.exit_code == 0, repr((res.output, res.stdout, res.stderr, res.exception))
+
+
+@pytest.mark.timeout(5 * 60)  # image pulling is very slow
+def test_file_outside_curdir(runner: CliRunner):
+    with tempfile.TemporaryDirectory() as dir:
+        os.mkdir(f'{dir}/working_dir')
+        os.chdir(f'{dir}/working_dir')
+        write_hello(f'{dir}/hello.txt')
+        write_script(dir, '/hello.txt')
+        res = runner.invoke(
+            cli.app, ['submit', '--wait', '--files', f'{dir}/hello.txt:/', '../test_job.py'], catch_exceptions=False
+        )
+        assert res.exit_code == 0, repr((res.output, res.stdout, res.stderr, res.exception))
+
+
+@pytest.mark.timeout(5 * 60)  # image pulling is very slow
+def test_dir_outside_curdir(runner: CliRunner):
+    with tempfile.TemporaryDirectory() as dir:
+        os.mkdir(f'{dir}/working_dir')
+        os.chdir(f'{dir}/working_dir')
+        write_hello(f'{dir}/hello1.txt')
+        write_hello(f'{dir}/hello2.txt')
+
+        dir_basename = os.path.basename(dir)
+        write_script(dir, f'/{dir_basename}/hello1.txt')
```

Comment:
Can you explain why the behavior of needing to use `basename` is different than in `test_dir_mount_in_child_dir_to_child_dir`?

Reply:
Yeah I think maybe there's still a bug here. The semantics of `/` that hailctl currently uses doesn't work well with symlinks.

In `test_dir_mount_in_child_dir_to_child_dir`:
```
--files child/:/child/
```
The trailing slash on the source, `child/`, indicates we want to work with the contents of child. The trailing slash on the destination, `/child/`, indicates we want the destination to be *inside* `/child/`. This means hello1.txt and hello2.txt are placed inside the folder `/child/`. To be clear: I deduced these rules from the code, I'm not married to these being the rules.

In contrast, in `test_dir_outside_curdir`:
```
--files {dir}/:/
```
This *should* have the same semantics as above but we cannot symlink to `/` (the OS already exists there) so the only option is to put it in `/the-basename`. I suspect the code just breaks currently. We should maybe prohibit `x/:/`.

Reply:
I think we should prohibit it if we're doing something special.